securedrop/requirements/securedrop-app-code-requirements.txt

```diff
@@ -4,6 +4,7 @@
 #
 #    pip-compile --output-file securedrop/requirements/securedrop-app-code-requirements.txt securedrop/requirements/securedrop-app-code-requirements.in
 #
+-e git+https://github.com/freedomofpress/python-gnupg.git@af4508260d6eaa2b05efc59f8253585b7be29ad3#egg=gnupg
```

Comment:
@kushaldas @msheiny: Regarding this `psutil` issue you both ran into on import of `python-gnupg`, can you confirm that the issue was due to a zombie process? That is, was the exception raised `psutil.ZombieProcess`?

Reply:
And if yes, my next question is: _why_ are there so many `gpg-agent` zombie processes on the system? Was this seen in local machine, in the dev container, CI or some combination of the above?

Making a temporary fork simply for debugging makes a lot of sense :+1:. That said, we shouldn't merge the Tor Browser functional testing feature branch with the fork of `python-gnupg`. Generally, we'll avoid creating a fork of libraries unless it becomes absolutely necessary.

Reply:
I did not see any zombie processes. It was trying to access `PID 2`.

Reply:
Ah, hm, was the `psutil` exception? `NoSuchProcess`?

Reply:
nah @redshiftzero -- im pretty certain it was `psutil.AccessDenied`

Reply:
Just to clarify, that exception is really easy to reproduce locally (at least on my linux grsec box):

```python
import psutil
p = psutil.Process(2)
p.name
```

result

```
[...]

AccessDenied: psutil.AccessDenied (pid=2)
```

Reply:
i totally agree that we shouldn't maintain a fork here though.. i was about to implement a fix and noticed that a [PR](https://github.com/isislovecruft/python-gnupg/pull/203) already sitting around (pulled in July 2017) had what we need. So I pulled in that commit and rebased on lastest master.  @isislovecruft 's repo seemed slammed with PRs (i'm totally not blaming them and I just did the math on us getting that upstreamed quickly and decided to fork).

Apologies for the lack of comments here on my behalf. I slung that change to @kushaldas out of band while we were having issues collaborating with the different branches and should have gone back in to clarify in comments.

Reply:
The only reason we need psutil in the tests is to start/stop the rqworker process on-demand. What if we edited the `_get_pid_from_file` logic to `cat` the PID from the static filepath if it exists? Then we can run just that logic with `sudo` and presumably sidestep the issue with needing to maintain a fork.

Reply:
@conorsh We don't need the `psutil` in this case, it is the `gnupg` code which is doing this call.

Reply:
Hi @kushaldas, I think you meant to ping @conorsch 🙂

Reply:
Howdy, @conorsh! Thank you kindly for the redirect. :smile:

Reply:
@conorsh Oops, sorry :)

Reply:
OK, thanks @msheiny. I was not able to reproduce this issue on macOS on Friday (well, I can artificially generate an exception by putting a breakpoint in the `GPGMeta. _find_agent` method and killing the process, which produces `psutil.NoSuchProcess`, but that's not really a reproducer). 

The reason I was asking is to understand if the broad exception handling over in https://github.com/isislovecruft/python-gnupg/pull/203 was necessary (or if a [different approach](https://github.com/redshiftzero/python-gnupg/commit/f25954f303cf4ca58eadaf128f642f4a4d90bb22) would be better), but it seems like the original PR is the right call given that you hit `psutil.AccessDenied`, and the user in the original issue on FreeBSD (https://github.com/isislovecruft/python-gnupg/issues/201) that was having this issue hit `psutil.ZombieProcess`

Reply:
Hi @redshiftzero! Thanks for alerting me about this issue via email! I get a lot of github notifications and I can't address all of them. Sorry for the delay!

I've merged that PR and a couple more fixes. If there's anything else SecureDrop needs, feel free to ping me. I have a bit of time in the next couple days to fix bugs and/or review patches.

Reply:
The upstream PR has been merged: https://github.com/isislovecruft/python-gnupg/pull/203 so we're in a better position now to include the new logic, rather than maintaining a fork. My initial understanding of the introduction of this new dependency was that it was required to get the TBB logic working; during sprint planning today, however, @kushaldas indicated that the problem is _not_ related to the TBB work, and can sometimes occur on `develop`. 

One possibility is that the issue only arises on grsecurity-patched workstations, given that CI seems relatively happy, I'm able to run `make test` on my workstation on the `develop` branch without error, on a grsecurity-patched workstation, so I'm inclined to say that's _not_ the case. Either way, I'd like to remove this change from the current PR and defer to a separate ticket, so reduce the scope of work we're reviewing and queuing up for a future release.

Reply:
awesome, thanks @isislovecruft!